### PR TITLE
Add convert processor

### LINKED ISSUE #8124: [Processors] Type coercion of event fields
As a user I would like to be able to coerce field values to a specific type. When using the dissect processor with numeric values I would like to have them reported as numbers rather than strings.

In this example the `process.pid` value will be reported as a string rather than a number.

```
filebeat.inputs:
  - type: log
    paths:
      - data.log
    processors:
      - dissect:
          tokenizer: "pid=%{process.pid}"
          field: "message"
          target_prefix: ""
```

The Logstash's dissect process supports coersion with [convert_datatype](https://www.elastic.co/guide/en/logstash/current/plugins-filters-dissect.html#plugins-filters-dissect-convert_datatype). Alternatively you could use a mutate filter to [coerce](https://www.elastic.co/guide/en/logstash/current/plugins-filters-mutate.html#plugins-filters-mutate-coerce).

Ingest node has a dedicated `convert` processor. https://www.elastic.co/guide/en/elasticsearch/reference/master/convert-processor.html

I think have a processor just for coercion would be good since you might want to use with with processors like `decode_json_fields` to correct the type of some value.

### PR DESCRIPTION
**Pull request by andrewkroh**
The `convert` processor converts a field in the event to a different type, such
as converting a string to an integer. For a full description of the processor's
capabilities see the add documentation.

Closes #8124

### CODE REVIEW

**Comment by ruflin** [comment]
jenkins, test this

File: libbeat/processors/convert/convert.go [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+				return event, annotateError(p.Tag, errors.Errorf("field [%v] is missing, cannot be converted to type [%v]", conv.From, conv.Type))
+			}
+			continue
+		}
+
+		if conv.Type > unset {
+			t, err := p.transformType(conv.Type, v)
+			if err != nil {
+				if !p.IgnoreFailure {
+					return event, annotateError(p.Tag, errors.Wrapf(err, "unable to convert field [%v] value [%v] to [%v]", conv.From, v, conv.Type))
```
Comment (ruflin) [inline libbeat/processors/convert/convert.go]: Right now on failure the event might be already modified by previous fields. Is this expected or should it reset to the original event? I think we do that in other processors. Same above with ignore missing.

**Comment by andrewkroh** [comment]
> is that it does not only convert but also rename and copy.

The [ingest node convert processor](https://www.elastic.co/guide/en/elasticsearch/reference/master/convert-processor.html) does a copy. Adding the ability to choose rename over copy is something I added. The reason I want the option to rename is to tie it in with the conversion such that the source field can be deleted iff the conversion is successful. This makes it possible to degrade gracefully if some of the data you are transforming isn't perfect.

For example if I'm trying to move `event_data.SourceIP` to `source.ip` but the value is `-` then the conversion will fail because the IP in invalid (which is great because it will prevent an ES mapping type exception later). In that failure case I want to retain the original field `event_data.SourceIP`. And if it's a valid IP then want to drop the `event_data.SourceIP` field.

Without the rename option you could accomplish this with a `convert` followed by conditional `drop_fields` for each field (that would be kind of tedious). You should not use a `rename` processor followed by `convert` because that could get you into a mapping exception if you end up writing say a `string` into a field that's mapped as `ip`.

**Comment by ruflin** [comment]
I guessed that you got inspirted by the other convert processors. I'm good with moving forward as you suggested.

Can you check my comment here? https://github.com/elastic/beats/pull/11686#discussion_r274305577

File: libbeat/processors/convert/convert.go (hunk 1 above)
Comment (andrewkroh) [reply libbeat/processors/convert/convert.go]: @ruflin I have finally addressed this and added a few tests to check that the event is not modified at all if there is a failure.

File: libbeat/processors/convert/convert.go [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+	defer resetValues(p.converted)
+
+	// Validate the conversions.
+	for i, conv := range p.Fields {
+		v, err := convertField(event, conv.From, conv.Type)
+		if err != nil {
+			switch cause := errors.Cause(err); cause {
+			case common.ErrKeyNotFound:
+				if !p.IgnoreMissing {
+					return event, annotateError(p.Tag, errors.Errorf("field [%v] is missing, cannot be converted to type [%v]", conv.From, conv.Type))
```
Comment (adriansr) [inline libbeat/processors/convert/convert.go]: Should this still return an error when ignore_failure is set? I've seen other processors with both flags wouldn't

File: libbeat/processors/convert/config.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		IgnoreMissing: false,
+		Mode:          copyMode,
+	}
+}
+
+type config struct {
+	Fields        []field `config:"fields" validate:"required"` // List of fields to convert.
+	Tag           string  `config:"tag"`                        // Processor ID for debug and metrics.
+	IgnoreMissing bool    `config:"ignore_missing"`             // Skip field when From field is missing.
+	IgnoreFailure bool    `config:"ignore_failure"`             // Ignore conversion errors.
```
[the remark this answers is not in the document]
Comment (andrewkroh) [reply libbeat/processors/convert/config.go]: Good point. I changed it to `fail_on_error`. I was modeling it directly off of the Ingest Node processor and neglected to notice this convention in Beats.

File: libbeat/processors/convert/convert.go (hunk 2 above)
Comment (andrewkroh) [reply libbeat/processors/convert/convert.go]: I added a check for this.

File: libbeat/processors/convert/convert.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+// cloneValue returns a shallow copy of a map. All other types are passed
+// through in the return. This should be used when making straight copies of
+// maps without doing any type conversions.
+func cloneValue(value interface{}) interface{} {
+	switch v := value.(type) {
+	case common.MapStr:
+		return v.Clone()
+	case map[string]interface{}:
+		return common.MapStr(v).Clone()
+	default:
```
Comment (adriansr) [inline libbeat/processors/convert/convert.go]: Here, @urso [suggests](https://github.com/elastic/beats/pull/11761#discussion_r280467032) adding a case for arrays too.

I [added](https://github.com/elastic/beats/pull/11761/commits/03adad0aaad4c54145dca2bf754ddc21cf0401cc) a simple case for []interface{}

Comment (andrewkroh) [reply libbeat/processors/convert/convert.go]: Thanks, I copied that in.

**Pull request merged**